### ISSUE TITLE
web3.eth.sign() methods failed: eth_sign requires 32 byte message hash

### ISSUE THREAD

**Issue by lbhsot** [ethereum/web3.js#4780]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

```
    const eth: any = await detectEthereumProvider();
    const web3 = new Web3(eth);
    await eth.request({
      method: 'wallet_requestPermissions',
      params: [{ eth_accounts: {} }],
    });
    const accounts = await web3.eth.getAccounts();
    if (accounts.length > 0) {
      const signature = await web3.eth.sign('hello world', accounts[0]);
      console.log(signature);
    }
```

The code will failed, error message: `eth_sign requires 32 byte message hash`

### Expected Behavior

I can sign use `web3.eth.sign` success.

### Steps to Reproduce

```
    const eth: any = await detectEthereumProvider();
    const web3 = new Web3(eth);
    await eth.request({
      method: 'wallet_requestPermissions',
      params: [{ eth_accounts: {} }],
    });
    const accounts = await web3.eth.getAccounts();
    if (accounts.length > 0) {
      const signature = await web3.eth.sign('hello world', accounts[0]);
      console.log(signature);
    }
```

### Web3.js Version

1.7.0

### Environment

- Operating System: 
- Browser: 
- Node.js Version: 
- NPM Version: 

### Anything Else?

I saw an existing issue: https://github.com/ChainSafe/web3.js/issues/4657

The issue is closed but bug still there. Anyone can help?

**Comment by nazarhussain** [comment]
@lbhsot Thanks for reaching back to us and notifying about the issue. I will investigate and get back to you.

**Comment by nazarhussain** [comment]
@lbhsot I quickly wrote a hardhat test to verify, and it seems to be working. 

```js
describe("4780 - Eth Sign", () => {
  it("sign an account", async () => {
    const accounts = await hre.web3.eth.getAccounts();
    const signature = await hre.web3.eth.sign('hello world', accounts[0]);

    console.log("%%%% web3: ", hre.web3.version);
    console.log("%%%% account:", accounts[0]);
    console.log("%%%% signature:", signature);

    assert.equal(signature.length > 0, true);
  });
});

```

The output is: 

```
%% web3:  1.7.0
%% account: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
%% signature: 0xa461f509887bd19e312c0c58467ce8ff8e300d3c1a90b608a760c5b80318eaf15fe57c96f9175d6cd4daad4663763baa7e78836e067d0163e9a2ccf2ff753f5b1b
```

Can you please verify what provider you are using? As at the end of flow the sign is actually been processed by provider.

**Comment by lbhsot** [comment]
@nazarhussain I use metamask as the provider. Is it means different provider will show different results?

**Comment by zoix** [comment]
Same issue here. As mentioned in #4657, `web3.eth.personal.sign` is also the alternative we are using as a workaround. It seems to work when the account is unlocked and setting  `null` or any string as the `password`.

As for `web3.eth.sign`, it throws the error described by @lbhsot.

---

**Update:**

This seems to work, but still not ideal:

`const signedMessage = await web3.eth.sign(web3.utils.sha3('Hello world'), address)`

**Comment by nazarhussain** [comment]
@lbhsot If wallet is not loaded or created within `web3` library then the provider is actually gonna handle the implementation of `sign`. There could be different behavior from different providers as I mentioned above it's working with hardhat. I will check with metamask to be sure.

**Comment by nazarhussain** [comment]
I investigated it further and can confirm it's an issue or behavior change in the Metamask. That particular error is thrown by the MetaMask app in the browser, later catched and returned by web3. I tried with Hardhat and Ganache providers and both works fine. 

The major difference is how different providers is handling `eth_sign` RPC method.

The Metamask [documented](https://docs.metamask.io/guide/signing-data.html#a-brief-history) it as. And that's the actual issue, it expect just a 32 byte message hash instead of any data. 

> the method eth_sign is an open-ended signing method that allows signing an arbitrary hash, 

The Eth RPC [documentation](https://eth.wiki/json-rpc/API) on the other hand, refers it as data with any number of bytes. 

> DATA, N Bytes - message to sign

Most of the providers are following this spec and so why working along web3js. Remember that `web3.js` is just passing the RPC call for `eth_sign` to the particular provider you are connected. The `web3.accounts.sign` 


@lbhost I suggest to take up this issue to the Metamask team to get their opinion on it.
